test/analysis/typeComplexityAnalysis.test.ts

```diff
@@ -0,0 +1,328 @@
+         *
+         * Anouther example, if we queried the 'Search' type with some string argument, the returned number of objects
+         * could be very large. Our algorithm will need to know what limit is set for the returned data (limit 100 search results
+         * for example) and then account for that response to caculate the complexity. That information is in the resolvers. We
+         * have no access to the resolvers.
+         *
+         * Some user configuration will be needed unless someone has bright ideas.
+         */
+        // ? type weigts are variable, not sure how to calculate this.
+        test('with lists', () => {
+            query = `
+            Query { 
+                human(id: 1) { 
+                    name, 
+                    friends { 
+                        name 
+                    } 
+                }
+            }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(7); // 1 Query + 1 human/character +  5 friends/character
+            query = `Query {reviews(episode: EMPIRE, first: 3) { stars, commentary } }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(4); // 1 Query + 3 reviews
+        });
+
+        test('with nested lists', () => {
+            query = `
+            query { 
+                human(id: 1) { 
+                    name, 
+                    friends(first: 5) { 
+                        name, 
+                        friends(first: 3){ 
+                            name 
+                        } 
+                    } 
+                }
+            }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(17); // 1 Query + 1 human/character +  (5 friends/character X 3 friends/characters)
+        });
+
+        test('accounting for __typename feild', () => {
+            query = `
+            query {
+                search(text: "an", first: 4) {
+                    __typename
+                    ... on Human {
+                        name
+                        homePlanet
+                    }
+                    ... on Droid {
+                        name
+                        primaryFunction
+                    }
+                }
+            }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(5); // 1 Query + 4 search results
+        });
+
+        // todo: directives @skip, @include and custom directives
+
+        // todo: error handling
+        // look into error handling for graphql. The only error I forsee is if the query is invalid in
+        // which case we want to pass the query along to the graphQL server to handle.
+        // What would that look like here? I think we should throw ar error from this function.
+        test('Throws an error if for a bad query', () => {
+            query = `Query { hello { hi } }`; // type doesn't exist
+            expect(getQueryTypeComplxity(query, typeWeights)).toThrow('Error');
```

Comment:
Minor but missing a 'e' in getQueryTypeComplxity and you can just use `.toThrow()` which checks if any error is thrown. 

I agree on this function throwing an error. A  question for the middleware then is do we want to pass the error down the middleware chain or send back a 400 status indicating a malformed request

Reply:
I was thinking we pass it onto to GraphQL to handle, so that thir error handling can respond with their error messages. We can discuss.

Reply:
If we use the `parse()` method and can respond to the client in the same way that GraphQL does then we can send the error back to the client. Otherwise calling next() will just let GraphQL handle this. 

- [ ] Let's create an issue to dive further into error handling